File: DoubtClass-24(LongestSubstringWithKUniqueCharacters-CloneListWithRandomPointers)/CloneListWithRandomPointers.java
/** This is the ListNode class definition

class ListNode {
	int data;
	ListNode next;
    ListNode random;

	ListNode (int data) {
        this.data = data;
        this.next = null;
        this.random = null;
    }
}
**/
class Solution {
	ListNode cloneTheLinkedList(ListNode head) {
		ListNode copy = null;
		ListNode a = head;
		ListNode b = null;
		HashMap<ListNode, ListNode> hashMap = new HashMap<ListNode, ListNode>();
		while(a != null) {
			ListNode temp = new ListNode(a.data);
			if(b == null) {
				copy = b = temp;
			}
			else {
				b.next = temp;
				b = b.next;
			}
			hashMap.put(a, temp);
			a = a.next;
		}
		a = head;
		b = copy;
		while(b != null) {
			b.random = hashMap.get(a.random);
			a = a.next;
			b = b.next;
		}
		return copy;
	}
}

File: DoubtClass-24(LongestSubstringWithKUniqueCharacters-CloneListWithRandomPointers)/CloneListWithRandomPointers.cpp
/* This is the ListNode class definition

class ListNode {
    public:
    int data;
    ListNode *next;
    ListNode* random;
    ListNode (int data) {
        this->data = data;
        this->next = NULL;
        this->random = NULL;
    }
};
*/

ListNode* cloneTheLinkedList (ListNode* head) {
    // add your logic here
	ListNode* a = head;
	ListNode* copy = NULL;
	ListNode* b = NULL;
	map<ListNode*, ListNode*> ump;
	while(a != NULL) {
		ListNode* temp = new ListNode(a->data);
		if(copy == NULL) {
			copy = b = temp;
		}
		else {
			b->next = temp;
			b = b->next;
		}
		ump[a] = temp;
		a = a->next;
	}
	b = copy;
	a = head;
	while(b != NULL) {
		b->random = ump[a->random];
		a = a->next;
		b = b->next;
	}
	return copy;
}